core/hashbuffer.go

```diff
@@ -0,0 +1,138 @@
+// It is not thread-safe.
+type hashBuffer struct {
+	// The data holds the hashes. The hashes are sequential, but also a
+	// circular buffer.
+	// The `head` points to the position of the latest hash.
+	// The parent, if present, is located 32 bytes back.
+	// [.., .., ..., head-2 , head-1, head, oldest, ... ]
+	data [hashBufferElems]common.Hash
+
+	head uint64 // index of hash for the head block
+
+	headNumber uint64 // The block number for head (the most recent block)
+	tailNumber uint64 // The block number for tail (the oldest block)
+}
+
+// newHashStorage creates a new storage with a header in it.
+// Since we take a header here, a hash storage can never be empty.
+// This makes things easier later on (in Set)
+func newHashStorage(header *types.Header) *hashBuffer {
+	return &hashBuffer{
+		headNumber: header.Number.Uint64(),
+		tailNumber: header.Number.Uint64(),
+		data:       [hashBufferElems]common.Hash{header.Hash()},
+	}
+}
+
+// Get locates the hash for the requested number
+func (hs *hashBuffer) Get(number uint64) (common.Hash, bool) {
+	if !hs.has(number) {
+		return common.Hash{}, false
+	}
+	distance := hs.headNumber - number
+	index := (hs.head + hashBufferElems - distance) % hashBufferElems
+	return hs.data[index], true
+}
+
+// has returns if the storage has a hash for the given number
+func (hs *hashBuffer) has(number uint64) bool {
+	return number <= hs.headNumber && number >= hs.tailNumber
+}
+
+// Contains checks if the hash at the given number matches the expected
+func (hs *hashBuffer) Contains(number uint64, expected common.Hash) bool {
+	if !hs.has(number) {
+		return false
+	}
+	distance := hs.headNumber - number
+	index := (hs.head + hashBufferElems - distance) % hashBufferElems
+	return hs.data[index] == expected
+}
+
+// Newest returns the most recent (number, hash) stored
+func (hs *hashBuffer) Newest() (uint64, common.Hash) {
+	return hs.headNumber, hs.data[hs.head]
+}
+
+// Oldest returns the oldest (number, hash) found
+func (hs *hashBuffer) Oldest() (uint64, common.Hash) {
+	distance := hs.headNumber - hs.tailNumber
+	index := (hs.head + hashBufferElems - distance) % hashBufferElems
+	return hs.tailNumber, hs.data[index]
+}
+
+// Set inserts a new header (hash) to the storage.
+// If
+// a) Header already exists, this is a no-op
+// b) Number is occupied by other header, the new header replaces it, and also
+// truncates any descendants
+//
+// If the new header does not have any ancestors, it replaces the entire storage.
+func (hs *hashBuffer) Set(header *types.Header) {
+	var (
+		number = header.Number.Uint64()
+		index  uint64
+		hash   = header.Hash()
+	)
+	if hs.Contains(number-1, header.ParentHash) {
+		if hs.headNumber >= number {
+			distance := hs.headNumber - number
+			index = (hs.head + hashBufferElems - distance) % hashBufferElems
+			if bytes.Compare(hs.data[index][:], hash[:]) == 0 {
+				return
```

Comment:
I'm unsure about this return. If sync fails, I do a rollback of 16 blocks, and reimport those again, then they will not wipe subsequent ones. Although it should not cause any issues, I'm wondering if it would be more predictable to wipe everything after the inserted header even it it's a noop. Might be easier to reason about if a Set surely sets the header as the head and leaves nothing in there above it.

Reply:
I think it should remain as is. If we re-add an existing header, there's no point intentionally removing data that might be useful. The first forking header will wipe it.